pdns/syncres.hh

```diff
@@ -339,7 +346,30 @@ public:
   };
 
   typedef std::unordered_map<DNSName, DecayingEwmaCollection> nsspeeds_t;
-  typedef map<ComboAddress, EDNSStatus> ednsstatus_t;
+  
+  struct EDNSStatus {
+    EDNSStatus(const ComboAddress &arg) : address(arg) {}
+    ComboAddress address;
+    time_t modeSetAt{0};
+    enum EDNSMode { UNKNOWN=0, EDNSOK=1, EDNSIGNORANT=2, NOEDNS=3 } mode{UNKNOWN};
+  };
+  
+  struct ednsstatus_t : public multi_index_container<EDNSStatus,
+                                indexed_by<
+                                  ordered_unique<member<EDNSStatus, ComboAddress, &EDNSStatus::address>>,
+                                  ordered_non_unique<member<EDNSStatus, time_t, &EDNSStatus::modeSetAt>>
+                                  >> {
+    void reset(nth_index<0>::type &index, iterator it) {
+      index.modify(it, [](EDNSStatus &s) { s.mode = EDNSStatus::EDNSMode::UNKNOWN; s.modeSetAt = 0; }); 
+    }
+    void setMode(nth_index<0>::type &index, iterator it, EDNSStatus::EDNSMode mode) {
+      index.modify(it, [mode](EDNSStatus &s) { s.mode = mode; });
```

Comment:
You could make the `mode` member `mutable`, since it is not involved in any index, making it possible to update it directly.